Geometry_Engine: HashArray method added for BoundingBox and Plane

pawelbaran opened the pull request:
### Issues addressed by this PR
<!-- Add reference(s) to issue(s) solved by this PR. Please use keyword Fixes/Closes as per https://help.github.com/articles/closing-issues-using-keywords/ -->

Closes #3538

<!-- Add short description of what has been fixed -->


### Test files
<!-- Link to test files to validate the proposed changes -->
On [SharePoint](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?id=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F02%5FPull%20Request%2FBHoM%2FBHoM%5FEngine%2FGeometry%5FEngine%2F%233539%2DHashArrayForBoundingBox&viewid=a5a93cba%2Dfcca%2D46cc%2Da31a%2D80b8d5cbfd7b&FolderCTID=0x0120008122C8891F89054B8ACED0196C70DFC4) - it is an updated version of the file originally uploaded [here](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?id=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F02%5FPull%20Request%2FBHoM%2FBHoM%5FEngine%2FGeometry%5FEngine%2F%232929%2DGeometryHash&viewid=a5a93cba%2Dfcca%2D46cc%2Da31a%2D80b8d5cbfd7b&FolderCTID=0x0120008122C8891F89054B8ACED0196C70DFC4).


### Changelog
<!-- Text to go into changelog if applicable -->
<!-- Please see https://github.com/BHoM/documentation/wiki/changelog for guidelines -->


### Additional comments
<!-- As required -->
The newly added method is an educated rewrite of `HashArray(Basis)`.

pawelbaran committed:
HashArray method added for BoundingBox

pawelbaran commented:
@BHoMBot check compliance

pawelbaran commented:
@BHoMBot check required

pawelbaran commented:
@BHoMBot check unit-tests

alelom commented:
The PR looks good to me. The only thing I see the description of the PR is lacking an issue and it's currently stating:
> The newly added method is an educated rewrite of `HashArray(Basis)`.

while this PR seems to simply add a method for `BoundingBox` and doing some minor refactoring expliciting the type of `var` variables.

Also, in order to test these methods, it would be good to have unit tests. Happy for you to add them in any way you want. I've checked the https://github.com/BHoM/DiffingTests_Prototypes but it seems we have not tests there for `HashArray` methods yet, I'd be happy for you to add some tests there if you want. At some point we really need the entirety of DiffingTests_prototypes in the https://github.com/BHoM/BHoM_Engine/tree/develop/.ci/unit-tests solution, so I would be happy for you to add your tests for this new method directly there, even in the absence of the rest of the DiffingTests_prototypes. At some point, when we get the new DevOps person in place, we should get to be able to automatically run everything via CI/CD, hopefully 😄

pawelbaran committed:
HashArray method added for Plane

pawelbaran committed:
GeometryHash UTs added

pawelbaran committed:
UTs updated

pawelbaran commented:
Thanks @alelom for the review, all good points. What I did:
- I dug up your old GH script used to test that feature and updated as needed (not sure how did the UTs disappear from `develop` btw)
- added `HashArray` overload for `Plane` (was also missing for some reason)
- raised https://github.com/BHoM/BHoM_Engine/issues/3540, which most likely requires your intervention unfortunately

...I hope the above concludes the scope of this PR, ready for re-review 👍

pawelbaran commented:
@BHoMBot check compliance
@BHoMBot check unit-tests

alelom reviewed (approved):
Thank you for your work @pawelbaran , looks good, and it makes sense with the open issue and the grasshopper scripts! Sorry for the delay, I was convinced I already had re-reviewed

pawelbaran commented:
@BHoMBot check ready-to-merge
